osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2023 Alliander N.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.opensmartgridplatform.adapter.protocol.dlms.infra.networking;
+
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import lombok.extern.slf4j.Slf4j;
+import org.bouncycastle.util.encoders.Hex;
+import org.opensmartgridplatform.dlms.DlmsPushNotification;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.AlarmTypeDto;
+
+@Slf4j
+public class Mx382AlarmDecoder extends AlarmDecoder {
+
+  private static final int NUMBER_OF_BYTES_FOR_ADDRESSING = 8;
+  private static final byte EVENT_NOTIFICATION_REQUEST = (byte) 0xC2;
+  private static final byte[] AMM_FORWARDED_ALARM_VERSION_0 =
+      new byte[] {
+        (byte) 0x128, (byte) 0x128, (byte) 0x128, (byte) 0x128, (byte) 0x128, (byte) 0x00
+      };
+  private static final byte[] ALARM_REGISTER_3 =
+      new byte[] {(byte) 0x00, (byte) 0x00, (byte) 0x61, (byte) 0x62, (byte) 0x02, (byte) 0xFF};
+  private static final byte[] DEVICE_ID_2_CLASS_ID = new byte[] {0x00, 0x01};
+  private static final byte VALUE_ATTRIBUTE_ID = 0x02;
+  private static final byte DSMR22_EQUIPMENT_IDENTIFIER_LENGTH = 0x10;
+  private static final byte DSMR22_NUMBER_OF_BYTES_FOR_LOGICAL_NAME = 0x06;
+  private static final byte ELEMENTS_IN_STRUCTURE = 0x03;
+
+  /** DLMS data types used in the SMR5 push notification */
+  private static final byte STRUCTURE = 0x02;
+
+  private static final byte OCTET_STRING = 0x09;
+  private static final byte DOUBLE_LONG_UNSIGNED = 0x06;
+
+  private final DlmsPushNotification.Builder builder = new DlmsPushNotification.Builder();
+
+  public DlmsPushNotification decodeMx382alarm(final InputStream inputStream)
+      throws UnrecognizedMessageDataException {
+
+    // Skip addressing
+    this.skip(inputStream, NUMBER_OF_BYTES_FOR_ADDRESSING);
+    log.debug("skipped {} bytes", NUMBER_OF_BYTES_FOR_ADDRESSING);
```

Comment:
The forwarded Mx382 alarm message has a header with byte[] {0, 1, 0, 103, 0, 102, 0, <apdu-length>}. I'm not sure what source wport 103 means, but it can be checked